spring-cloud-dataflow-module-deployers/spring-cloud-dataflow-module-deployer-kubernetes/src/main/java/org/springframework/cloud/dataflow/module/deployer/kubernetes/KubernetesModuleDeployer.java

```diff
@@ -0,0 +1,202 @@
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.cloud.dataflow.core.ModuleDeploymentId;
+import org.springframework.cloud.dataflow.core.ModuleDeploymentRequest;
+import org.springframework.cloud.dataflow.module.ModuleStatus;
+import org.springframework.cloud.dataflow.module.deployer.ModuleDeployer;
+
+import io.fabric8.kubernetes.api.model.*;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClientException;
+
+
+/**
+ * Implements a deployer for the Google Kubernetes project. 
+ * 
+ * @author Florian Rosenberg
+ */
+public class KubernetesModuleDeployer implements ModuleDeployer {
+	
+	protected static final String SCSM_GROUP_KEY = "scsm-group";
+	protected static final String SCSM_LABEL_KEY = "scsm-label";
+	private static final String SCSM_EXTENSION = "scsm-extension";
+	private static final String SCSM_VERSION = "scsm-version";
+	private static final String SCSM_GROUP_ID = "scsm-groupId";
+	private static final String SCSM_ARTIFACT_ID = "scsm-artifactId";
+	private static final String SPRING_MARKER_VALUE = "scsm-module";
+	private static final String SPRING_MARKER_KEY = "role";
+	private static final String PORT_KEY = "port";
+	private static final String SERVER_PORT_KEY = "server.port";
+
+	private final Logger logger = LoggerFactory.getLogger(this.getClass());
+	
+	@Autowired
+	private KubernetesClient kubernetesClient;
+	
+	@Autowired
+	private ContainerFactory containerFactory;
+	
+	private KubernetesModuleDeployerProperties properties;
+
+	public KubernetesModuleDeployer(
+			KubernetesModuleDeployerProperties properties) {
+		this.properties = properties;
+	}
+
+	public ModuleDeploymentId deploy(ModuleDeploymentRequest request) {
+		ModuleDeploymentId id = ModuleDeploymentId.fromModuleDefinition(request.getDefinition());
+
+		logger.debug("Deploying module: {}", createKubernetesName(id));
+
+		int externalPort = 8080;
+		// we also create a service in case have an '--port' or '--server.port' argument (source or sink)
+		Map<String, String> parameters = request.getDefinition().getParameters();
+		if (parameters.containsKey(PORT_KEY)) {
+			externalPort = Integer.valueOf(parameters.get(PORT_KEY));
+			createService(id, request, externalPort);
+		} else if (parameters.containsKey(SERVER_PORT_KEY)) {
+			externalPort = Integer.valueOf(parameters.get(SERVER_PORT_KEY));		
+			createService(id, request, externalPort);
+		}		
+		createReplicationController(id, request, externalPort);				
+		return id;	
+	}
+
+	@Override
+	public void undeploy(ModuleDeploymentId id) {
+		String name = createKubernetesName(id);
+		logger.debug("Undeploying module: {}", name);
+
+		try {
+			kubernetesClient.replicationControllers().withName(name).delete();
+			kubernetesClient.services().withName(name).delete();
```

Comment:
You mean it would not fall in the catch block below (which would be bad)?
Let's make sure this is indeed the case (with a test) or try to come up with a less aggressive approach. From comments above, this needs to change from `withName()` to "withLabels()" approach anyway, which may indeed silently do nothing when there is no match